src/collector/api/v1/handler/prometheus/remote/write.go

```diff
@@ -0,0 +1,113 @@
+package remote
+
+import (
+	"fmt"
+	"io/ioutil"
+	"math"
+	"net/http"
+
+	"github.com/m3db/m3/src/collector/reporter"
+	"github.com/m3db/m3/src/metrics/metric/id"
+	"github.com/m3db/m3/src/query/models"
+	"github.com/m3db/m3/src/query/storage"
+	"github.com/m3db/m3/src/x/serialize"
+	"github.com/m3db/m3x/instrument"
+
+	"github.com/gogo/protobuf/proto"
+	"github.com/golang/snappy"
+	"github.com/prometheus/prometheus/prompb"
+)
+
+const (
+	WriteURL         = "/api/v1/write"
+	ReportHTTPMethod = http.MethodPost
+)
+
+type PromWriteHandler struct {
+	reporter       reporter.Reporter
+	encoderPool    serialize.TagEncoderPool
+	decoderPool    serialize.TagDecoderPool
+	instrumentOpts instrument.Options
+}
+
+func NewPromWriteHandler(
+	reporter reporter.Reporter,
+	encoderPool serialize.TagEncoderPool,
+	decoderPool serialize.TagDecoderPool,
+	instrumentOpts instrument.Options,
+) *PromWriteHandler {
+	return &PromWriteHandler{
+		reporter:       reporter,
+		encoderPool:    encoderPool,
+		decoderPool:    decoderPool,
+		instrumentOpts: instrumentOpts,
+	}
+}
+
+func (h *PromWriteHandler) ServeHTTP(w http.ResponseWriter, r *http.Request) {
+	compressed, err := ioutil.ReadAll(r.Body)
+	if err != nil {
+		http.Error(w, err.Error(), http.StatusInternalServerError)
+		return
+	}
+
+	data, err := snappy.Decode(nil, compressed)
+	if err != nil {
+		http.Error(w, err.Error(), http.StatusBadRequest)
+		return
+	}
+
+	var req prompb.WriteRequest
+	err = proto.Unmarshal(data, &req)
+	if err != nil {
+		http.Error(w, err.Error(), http.StatusBadRequest)
+		return
+	}
+
+	for _, v := range req.Timeseries {
+		id, err := h.newMetricID(v.Labels)
+		if err != nil {
+			http.Error(w, err.Error(), http.StatusInternalServerError)
+			return
+		}
+		for _, s := range v.Samples {
+			if math.IsNaN(s.Value) {
+				continue
+			}
+			err = h.reporter.ReportGauge(id, s.Value)
+			if err != nil {
+				http.Error(w, err.Error(), http.StatusInternalServerError)
+				return
+			}
+		}
+	}
+
+	w.WriteHeader(http.StatusOK)
+}
+
+func (h *PromWriteHandler) newMetricID(l []*prompb.Label) (id.ID, error) {
```

Comment:
We should probably move this logic to a central place so we don't duplicate this with the existing remote write logic. 

Perhaps we can put it in the `src/cmd/coordinator/downsample` package and export it?

Reply:
Yeah, I will refer to your comments to make changes.